### PR TITLE
Make find panel use standard combo boxes for history management

### PR DESCRIPTION
**Pull request by ThatsJustCheesy**
As mentioned on the [laundry list](https://github.com/textmate/textmate/wiki/Hacking-on-the-Code), the Find panel now uses standard combo box controls for history management, and OakPasteboardSeleector, which previousy filled this role, has been completely removed.

Also fixes a small longstanding issue of history items drawing a white background behind them when syntax colouring is used; it was _just barely_ visible in the "Find History" panel before, but with 10.14's translucent combo box dropdowns, it definitely stands out. Thus, there is a new option on OakSyntaxFormatter that tells it to omit the background from styled text, and it's used in both the Find panel and the Find History panel.

I hereby dedicate the contents of these patches to the public domain.

### CODE REVIEW

File: Frameworks/OakAppKit/src/OakUIConstructionFunctions.mm [hunk 1]
```diff
@@ -202,9 +202,11 @@ - (NSSize)intrinsicContentSize
...
 {
-	_style = aStyle;
-	[self updateBackgroundStyle];
-	self.needsDisplay = YES;
+	dispatch_async(dispatch_get_main_queue(), ^{
+		self->_style = aStyle;
+		[self updateBackgroundStyle];
+		self.needsDisplay = YES;
+	});
 }
```
[the remark this answers is not in the document]
Comment (ThatsJustCheesy) [reply Frameworks/OakAppKit/src/OakUIConstructionFunctions.mm]: Sure, here: https://pastebin.com/n7Nw27i6

**Comment by rdwampler** [comment]
@ThatsJustCheesy, if you remove `self.needsDisplay = YES` does that prevent the crash? 
I'm not seeing this crash, but looking through the code base, we do have similar places where we update the UI after observing `NSUserDefaultsDidChangeNotification`. Just wondering if we need to audit those call sites as well.

**Comment by sorbits** [comment]
> Just wondering if we need to audit those call sites as well.

Was thinking the same thing: might want to use `addObserverForName:object:queue:usingBlock:` with `mainQueue` for all observers of `NSUserDefaultsDidChangeNotification`.

In theory I guess it is not unlikely that a background thread does something that triggers a change of user defaults, although it is not something I have seen causes crashes before, but the stack dump from @ThatsJustCheesy doesn’t make it obvious that thread 6 triggered this notification (AFACT).

**Comment by sorbits** [comment]
@ThatsJustCheesy For the PR, thanks for taking the time to address this long standing issue :)

There are however some minor issues; for not drawing the background color: We still need to draw it when the background is changed from non-default, for example if you enter `${1:/foo}` in the replacement field with the Mac Classic theme, the invalid `foo` was previously white on red but is now white on white.

The theme does have a property to get the “global” background color, it will need to compare against this when deciding if background color should be set or not.

And I would appreciate this as a separate commit, as it’s not really related to the combo boxes, also, I do not think we need a property to disable background colors if it only removes the root/global background.

**Comment by ThatsJustCheesy** [comment]
> @ThatsJustCheesy, if you remove `self.needsDisplay = YES` does that prevent the crash?

Still crashes with the same stack dump. The crash is an exception being thrown in the call to `-addSubview:positioned:relativeTo:`, within `-updateBackgroundStyle`.

**Comment by sorbits** [comment]
## Issue

1. Search for regular expression `foo|bar` — close find dialog.
2. Uncheck regular expression and search for plain string `foo` — close find dialog.
3. Open find dialog, use combobox to reselect the `foo|bar` search string.

## Expected

Regular expression checkbox becomes checked.

## Actual

The regular expression checkbox remains unchecked.

**Comment by ThatsJustCheesy** [comment]
Ah, okay—did not know that was a feature, but now it seems quite obvious! I'll make sure that works.

**Comment by ThatsJustCheesy** [comment]
I've gone and mucked up my history after attempting to integrate the new changes on top of mine. I'll figure that out likely in a couple days.

Anyway, I've got it working nicely, aside from an occasional issue in which no history appears after adding a bunch of new items or something; I'll look into it. Screwy git history aside, the code is cleaner (and works better) now ;) .

(P.S. Do you perchance know of any good tutorials on how to use git effectively? I'm still a novice and it drives me bonkers)

**Comment by sorbits** [comment]
Noticed another issue: The auto-sizing combo-boxes do not have any max width.

So if e.g. you select 50+ lines of code, press ⌘E, and then open the find dialog, the large combo-box will push out the other controls of the window.

Prior to this PR the boxes were limited to (I think) 12 lines fo text.

Related: Prior to this PR you could use arrow up/down to move the insertion point in the find/replace fields, only on the last line, would arrow down open the history list.

With `NSComboBox` arrow down can no longer be used to move down the insertion point. Unsure how easy it would be to fix though.

**Comment by kfjelsted** [comment]
> Please make sure that this doesn't
cause any accessibility problems with VoiceOver. I just used find yesterday and other than the edit menu not saying that find is actually find and replace it worked great!
-Kevin

**Comment by sorbits** [comment]
@ThatsJustCheesy I merged the `OakSyntaxFormatter` fix as 9baff6d7e1e535909564a7ff47f80b4b35a4a146.

You should be able to do the `git fetch …` followed by `rebase -i` mentioned above, and then just `drop` your local commit (I made minor syntactic change).

Also, I found one instance where `NSUserDefaults` could be updated from a background thread which is fixed in 045fd33497eca97670a6b0c7483202ee483c2568 — I suggest you drop the “Make sure UI is updated on main thread” commit and test if you are still seeing any issues (locally I added an observer to user defaults with an assert of `NSThread.isMainThread` and the assertion hasn’t been broken yet).

**Comment by ThatsJustCheesy** [comment]
This took _waay_ too much fiddling, but aside from a couple minor graphical issues that frankly I don't think will bother anyone:
* When you switch off Regular Expression and open replace history, the currently selected item appears formatted in the list until a new selection is made; and
* Sometimes (again when switching between normal and regex mode), opening the list and pressing an arrow key deselects the currently selected item, due to some weird interaction

It otherwise appears bug-free and works like a charm. I've tested VoiceOver too, no problems there.

@sorbits Thank you for your patience with this and your hard work on this app. It's the sort of attention to detail that is spread through the entire product and that I had to replicate here that solidifies TextMate as my editor of choice and keeps me fleeing from competitors. Since I'm a student, this is my way of giving back for now. :)

**Comment by ThatsJustCheesy** [comment]
Whoops, fixed!

I had to reset the selection on the combo box whenever the user selects a new value so that e.g. arrow key navigation works at all. As of f80d11c this is only done when the combo box's selection list is open.

**Comment by ThatsJustCheesy** [comment]
I'm going to close this now. No hard feelings, and looking back the app is probably better off without this change.

I'm still a happy user of TextMate, thanks for your continued work :)

Ian

**Comment by sorbits** [comment]
Sorry about my silence on this one, and thanks for your comment: I had 3 issues noted for this PR:

1. Inability to delete items when the menu was up (via <kbd>⌃D</kbd> or <kbd>⌫</kbd>)
2. Arrow down always opening the menu, even when at the first line of multi-line search strings (rather than go down to next line)
3. Performance issue when changing selection of menu

The last one was to blame on the implementation of `OakPasteboard` (and the way it, until just a few days ago, used CoreData), therefore that was really my issue to fix, so I didn’t feel I should bring up the other two issues before I had at least fixed the third :)

It would still be nice to see TextMate use standard controls for the history, but I am not sure it will be possible to fix the two remaining issues, and without that, I do think it will be a worse user experience.

**Pull request closed without merging**